File: cpp/examples/coredevice_configuration.cpp
// Jackson Coxson
// Device configuration over the CoreDevice configuration service: the
// appearance and accessibility toggles Xcode drives.

#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <string>

#include <idevice++/configuration_service.hpp>
#include <idevice++/core_device_proxy.hpp>
#include <idevice++/ffi.hpp>
#include <idevice++/provider.hpp>
#include <idevice++/rsd.hpp>
#include <idevice++/usbmuxd.hpp>

using namespace IdeviceFFI;

[[noreturn]]
static void die(const char* msg, const FfiError& e) {
    std::cerr << msg << ": " << e.message << " (" << e.code << ")\n";
    std::exit(1);
}

static void usage(const char* argv0) {
    std::cerr << "Usage:\n"
              << "  " << argv0 << " get\n"
              << "  " << argv0 << " style <light|dark>\n"
              << "  " << argv0 << " opacity <0.0-1.0>\n"
              << "  " << argv0 << " text-size <name>\n"
              << "  " << argv0 << " color-filter <off|preset> [intensity]\n"
              << "  " << argv0 << " reduce-motion <on|off>\n"
              << "  " << argv0 << " reduce-transparency <on|off>\n"
              << "  " << argv0 << " increase-contrast <on|off>\n"
              << "  " << argv0 << " show-borders <on|off>\n";
}

static bool on_off(const std::string& value) { return value == "on" || value == "1"; }

int         main(int argc, char** argv) {
    if (argc < 2) {
        usage(argv[0]);
        return 2;
    }
    const std::string cmd = argv[1];
    const std::string arg = argc >= 3 ? argv[2] : std::string();

    auto              mux     = UsbmuxdConnection::default_new(0).expect("failed to reach usbmuxd");
    auto              devices = mux.get_devices().expect("failed to list devices");
    if (devices.empty()) {
        std::cerr << "no devices connected\n";
        return 1;
    }
    auto& dev    = devices[0];
    auto  udid   = dev.get_udid();
    auto  mux_id = dev.get_id();
    if (udid.is_none() || mux_id.is_none()) {
        std::cerr << "device has no UDID or mux id\n";
        return 1;
    }

    auto provider = Provider::usbmuxd_new(UsbmuxdAddr::default_new(),
                                          0,
                                          udid.unwrap(),
                                          mux_id.unwrap(),
                                          "coredevice-configuration")
                        .expect("failed to create provider");

    auto cdp = CoreDeviceProxy::connect(provider).unwrap_or_else(
        [](FfiError e) -> CoreDeviceProxy { die("failed to connect CoreDeviceProxy", e); });
    auto rsd_port = cdp.get_server_rsd_port().unwrap_or_else(
        [](FfiError e) -> uint16_t { die("failed to get the RSD port", e); });
    auto adapter = std::move(cdp).create_tcp_adapter().expect("failed to create the tunnel");
    auto stream  = adapter.connect(rsd_port).expect("failed to connect the RSD stream");
    auto rsd     = RsdHandshake::from_socket(std::move(stream)).expect("failed the RSD handshake");

    auto config  = ConfigurationService::connect_rsd(adapter, rsd)
                      .unwrap_or_else([](FfiError e) -> ConfigurationService {
                          die("failed to connect ConfigurationService", e);
                      });

    if (cmd == "get") {
        auto style = config.get_user_interface_style().unwrap_or_else(
            [](FfiError e) -> UserInterfaceStyle { die("get_user_interface_style failed", e); });
        std::cout << "style: " << (style == UserInterfaceStyle::Dark ? "dark" : "light") << "\n";

        auto size = config.get_device_text_size().unwrap_or_else(
            [](FfiError e) -> std::string { die("get_device_text_size failed", e); });
        std::cout << "text size: " << size << "\n";

        auto filter = config.get_color_filter().unwrap_or_else(
            [](FfiError e) -> ColorFilter { die("get_color_filter failed", e); });
        std::cout << "color filter: " << (filter.enabled ? "on" : "off") << " type="
                  << (filter.filter_type.is_some() ? filter.filter_type.unwrap()
                                                   : std::string("<none>"));
        if (filter.intensity.is_some()) {
            std::cout << " intensity=" << filter.intensity.unwrap();
        }
        std::cout << "\n";

        auto motion = config.get_reduce_motion().unwrap_or_else(
            [](FfiError e) -> bool { die("get_reduce_motion failed", e); });
        std::cout << "reduce motion: " << (motion ? "on" : "off") << "\n";

        auto transparency = config.get_reduce_transparency().unwrap_or_else(
            [](FfiError e) -> bool { die("get_reduce_transparency failed", e); });
        std::cout << "reduce transparency: " << (transparency ? "on" : "off") << "\n";

        auto borders = config.get_show_borders().unwrap_or_else(
            [](FfiError e) -> bool { die("get_show_borders failed", e); });
        std::cout << "show borders: " << (borders ? "on" : "off") << "\n";
        return 0;
    }

    if (argc < 3) {
        usage(argv[0]);
        return 2;
    }

    auto check = [](Result<void, FfiError> r) {
        if (r.is_err()) {
            die("the device rejected the change", r.unwrap_err());
        }
    };

    if (cmd == "style") {
        check(config.set_user_interface_style(arg == "dark" ? UserInterfaceStyle::Dark
                                                            : UserInterfaceStyle::Light));
    } else if (cmd == "opacity") {
        check(config.set_liquid_glass_opacity(std::stof(arg)));
    } else if (cmd == "text-size") {
        check(config.set_device_text_size(arg));
    } else if (cmd == "color-filter") {
        if (arg == "off") {
            check(config.set_color_filter(false, None, None));
        } else {
            Option<float> intensity;
            if (argc >= 4) {
                intensity = std::stof(argv[3]);
            }
            check(config.set_color_filter(true, arg, intensity));
        }
    } else if (cmd == "reduce-motion") {
        check(config.set_reduce_motion(on_off(arg)));
    } else if (cmd == "reduce-transparency") {
        check(config.set_reduce_transparency(on_off(arg)));
    } else if (cmd == "increase-contrast") {
        check(config.set_increase_contrast(on_off(arg)));
    } else if (cmd == "show-borders") {
        check(config.set_show_borders(on_off(arg)));
    } else {
        usage(argv[0]);
        return 2;
    }

    std::cout << "ok\n";
    return 0;
}

File: tools/src/coredevice_configuration.rs
// Jackson Coxson

use idevice::{
    IdeviceService, RsdService,
    core_device::{ConfigurationServiceClient, UserInterfaceStyle},
    core_device_proxy::CoreDeviceProxy,
    provider::IdeviceProvider,
    rsd::RsdHandshake,
};
use jkcli::{CollectedArguments, JkArgument, JkCommand, JkFlag};

pub fn register() -> JkCommand {
    let toggle = |help: &'static str| {
        JkCommand::new()
            .help(help)
            .with_argument(JkArgument::new().with_help("on or off; omit to read the current value"))
    };

    JkCommand::new()
        .help("Read and write appearance and accessibility knobs over CoreDevice")
        .with_subcommand(
            "appearance",
            JkCommand::new()
                .help("Get or set the light/dark appearance")
                .with_argument(
                    JkArgument::new().with_help("dark or light; omit to read the current style"),
                ),
        )
        .with_subcommand(
            "opacity",
            JkCommand::new()
                .help("Set the system liquid-glass opacity")
                .with_argument(
                    JkArgument::new()
                        .with_help("opacity, 0.0 to 1.0")
                        .required(true),
                ),
        )
        .with_subcommand(
            "color-filter",
            JkCommand::new()
                .help("Get or set the accessibility color filter")
                .with_argument(JkArgument::new().with_help("on or off; omit to read the state"))
                .with_argument(JkArgument::new().with_help("filter preset, e.g. Protanopia"))
                .with_argument(JkArgument::new().with_help("intensity, 0.0 to 1.0")),
        )
        .with_subcommand(
            "text-size",
            JkCommand::new()
                .help("Get or set the dynamic type size")
                .with_argument(
                    JkArgument::new().with_help("size name, e.g. large; omit to read it"),
                ),
        )
        .with_subcommand("reduce-motion", toggle("Get or set Reduce Motion"))
        .with_subcommand(
            "increase-contrast",
            JkCommand::new()
                .help("Set Increase Contrast (the device exposes no getter)")
                .with_argument(JkArgument::new().with_help("on or off").required(true)),
        )
        .with_subcommand(
            "show-borders",
            toggle("Get or set the layout-debug borders overlay"),
        )
        .with_subcommand(
            "reduce-transparency",
            toggle("Get or set Reduce Transparency"),
        )
        .with_flag(JkFlag::new("raw").with_help("Print raw values instead of a friendly summary"))
        .subcommand_required(true)
}

pub async fn main(arguments: &CollectedArguments, provider: Box<dyn IdeviceProvider>) {
    let (sub_name, sub_args) = arguments
        .first_subcommand()
        .expect("no subcommand passed, pass -h for help");
    let mut sub_args = sub_args.clone();

    let proxy = CoreDeviceProxy::connect(&*provider)
        .await
        .expect("no core device proxy");
    let rsd_port = proxy.tunnel_info().server_rsd_port;
    let adapter = proxy.create_software_tunnel().expect("no software tunnel");
    let mut adapter = adapter.to_async_handle();
    let stream = adapter.connect(rsd_port).await.expect("no RSD connect");
    let mut handshake = RsdHandshake::new(stream).await.unwrap();

    let mut client = ConfigurationServiceClient::connect_rsd(&mut adapter, &mut handshake)
        .await
        .expect("no configuration service");

    match sub_name.as_str() {
        "appearance" => match sub_args.next_argument::<String>().as_deref() {
            None => {
                let style = client
                    .get_user_interface_style()
                    .await
                    .expect("failed to read the appearance");
                println!("{}", style.as_str());
            }
            Some("dark") => set_appearance(&mut client, UserInterfaceStyle::Dark).await,
            Some("light") => set_appearance(&mut client, UserInterfaceStyle::Light).await,
            Some(other) => eprintln!("appearance must be 'dark' or 'light', got {other:?}"),
        },
        "opacity" => {
            let opacity: f32 = sub_args.next_argument().expect("opacity required");
            client
                .set_liquid_glass_opacity(opacity)
                .await
                .expect("failed to set the opacity");
            println!("opacity set to {opacity}");
        }
        "color-filter" => match sub_args.next_argument::<String>() {
            None => {
                let filter = client
                    .get_color_filter()
                    .await
                    .expect("failed to read the color filter");
                println!("{filter:#?}");
            }
            Some(state) => {
                let Some(enabled) = parse_bool(&state) else {
                    eprintln!("expected 'on' or 'off', got {state:?}");
                    return;
                };
                let filter_type = sub_args.next_argument::<String>();
                let intensity = sub_args.next_argument::<f32>();
                client
                    .set_color_filter(enabled, filter_type.as_deref(), intensity)
                    .await
                    .expect("failed to set the color filter");
                println!(
                    "color filter {}",
                    if enabled { "enabled" } else { "disabled" }
                );
            }
        },
        "text-size" => match sub_args.next_argument::<String>() {
            None => {
                let size = client
                    .get_device_text_size()
                    .await
                    .expect("failed to read the text size");
                println!("{size}");
            }
            Some(size) => {
                client
                    .set_device_text_size(&size)
                    .await
                    .expect("failed to set the text size");
                println!("text size set to {size}");
            }
        },
        "reduce-motion" => match sub_args.next_argument::<String>() {
            None => println!(
                "{}",
                client
                    .get_reduce_motion()
                    .await
                    .expect("failed to read reduce motion")
            ),
            Some(state) => {
                let Some(enabled) = parse_bool(&state) else {
                    eprintln!("expected 'on' or 'off', got {state:?}");
                    return;
                };
                client
                    .set_reduce_motion(enabled)
                    .await
                    .expect("failed to set reduce motion");
                println!("reduce motion {enabled}");
            }
        },
        "increase-contrast" => {
            let state: String = sub_args.next_argument().expect("on or off required");
            let Some(enabled) = parse_bool(&state) else {
                eprintln!("expected 'on' or 'off', got {state:?}");
                return;
            };
            client
                .set_increase_contrast(enabled)
                .await
                .expect("failed to set increase contrast");
            println!("increase contrast {enabled}");
        }
        "show-borders" => match sub_args.next_argument::<String>() {
            None => println!(
                "{}",
                client
                    .get_show_borders()
                    .await
                    .expect("failed to read show borders")
            ),
            Some(state) => {
                let Some(enabled) = parse_bool(&state) else {
                    eprintln!("expected 'on' or 'off', got {state:?}");
                    return;
                };
                client
                    .set_show_borders(enabled)
                    .await
                    .expect("failed to set show borders");
                println!("show borders {enabled}");
            }
        },
        "reduce-transparency" => match sub_args.next_argument::<String>() {
            None => println!(
                "{}",
                client
                    .get_reduce_transparency()
                    .await
                    .expect("failed to read reduce transparency")
            ),
            Some(state) => {
                let Some(enabled) = parse_bool(&state) else {
                    eprintln!("expected 'on' or 'off', got {state:?}");
                    return;
                };
                client
                    .set_reduce_transparency(enabled)
                    .await
                    .expect("failed to set reduce transparency");
                println!("reduce transparency {enabled}");
            }
        },
        _ => unreachable!(),
    }
}

async fn set_appearance(
    client: &mut ConfigurationServiceClient<Box<dyn idevice::ReadWrite>>,
    style: UserInterfaceStyle,
) {
    client
        .set_user_interface_style(style)
        .await
        .expect("failed to set the appearance");
    println!("appearance set to {}", style.as_str());
}

fn parse_bool(value: &str) -> Option<bool> {
    match value {
        "on" | "true" | "1" | "yes" => Some(true),
        "off" | "false" | "0" | "no" => Some(false),
        _ => None,
    }
}
